File: limablu_mobile/lib/services/allowance.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

const _kAllowanceKey = 'allowance_state_v1';

/// Serializable snapshot of the user’s allowance state.
class AllowanceState {
  final double monthlyAllowanceKwh;
  final double remainingKwh;
  final bool rollover;
  final DateTime lastReset; // start of current period (month)

  AllowanceState({
    required this.monthlyAllowanceKwh,
    required this.remainingKwh,
    required this.rollover,
    required this.lastReset,
  });

  AllowanceState copyWith({
    double? monthlyAllowanceKwh,
    double? remainingKwh,
    bool? rollover,
    DateTime? lastReset,
  }) {
    return AllowanceState(
      monthlyAllowanceKwh: monthlyAllowanceKwh ?? this.monthlyAllowanceKwh,
      remainingKwh: remainingKwh ?? this.remainingKwh,
      rollover: rollover ?? this.rollover,
      lastReset: lastReset ?? this.lastReset,
    );
  }

  Map<String, dynamic> toJson() => {
        'monthlyAllowanceKwh': monthlyAllowanceKwh,
        'remainingKwh': remainingKwh,
        'rollover': rollover,
        'lastReset': lastReset.toIso8601String(),
      };

  static AllowanceState fromJson(Map<String, dynamic> m) => AllowanceState(
        monthlyAllowanceKwh: (m['monthlyAllowanceKwh'] as num).toDouble(),
        remainingKwh: (m['remainingKwh'] as num).toDouble(),
        rollover: m['rollover'] as bool,
        lastReset: DateTime.parse(m['lastReset'] as String),
      );
}

class AllowanceService {
  SharedPreferences? _sp;
  AllowanceState? _state;

  /// Expose current state to the UI.
  AllowanceState? get state => _state;

  /// Load from SharedPreferences, or create a default.
  Future<void> load() async {
    _sp = await SharedPreferences.getInstance();
    final raw = _sp!.getString(_kAllowanceKey);
    if (raw != null) {
      final json = jsonDecode(raw) as Map<String, dynamic>;
      _state = AllowanceState.fromJson(json);
    } else {
      // sensible defaults for a brand-new user
      final now = DateTime.now();
      final startOfMonth = DateTime(now.year, now.month, 1);
      _state = AllowanceState(
        monthlyAllowanceKwh: 100,
        remainingKwh: 100,
        rollover: true,
        lastReset: startOfMonth,
      );
      await _save();
    }
  }

  /// If we’ve crossed into a new month, reset period.
  Future<void> ensureCurrentPeriod() async {
    if (_state == null) return;
    final now = DateTime.now();
    final currentPeriod = DateTime(now.year, now.month, 1);
    final lastPeriod = DateTime(_state!.lastReset.year, _state!.lastReset.month, 1);

    if (currentPeriod.isAfter(lastPeriod)) {
      // new month
      double newRemaining = _state!.monthlyAllowanceKwh;
      if (_state!.rollover) {
        newRemaining += _state!.remainingKwh; // carryover any unused
      }
      _state = _state!.copyWith(
        remainingKwh: newRemaining,
        lastReset: currentPeriod,
      );
      await _save();
    }
  }

  /// Update plan settings (and optionally mark as newly selected plan).
  Future<void> setPlan({
    required double monthlyAllowanceKwh,
    required bool rollover,
    bool newPlan = false,
  }) async {
    if (_state == null) await load();
    final s = _state!;
    // If selecting a brand-new plan, reset remaining to the plan amount (fresh start).
    final newRemaining = newPlan ? monthlyAllowanceKwh : s.remainingKwh;

    _state = s.copyWith(
      monthlyAllowanceKwh: monthlyAllowanceKwh,
      rollover: rollover,
      remainingKwh: newRemaining,
    );
    await _save();
  }

  /// Deduct used allowance and persist. Returns how much was actually deducted.
  Future<double> consume(double kwh) async {
    if (_state == null) await load();
    final s = _state!;
    final double deduct = kwh.clamp(0.0, s.remainingKwh);
    _state = s.copyWith(remainingKwh: s.remainingKwh - deduct);
    await _save();
    return deduct;
  }

  Future<void> _save() async {
    if (_sp == null) _sp = await SharedPreferences.getInstance();
    await _sp!.setString(_kAllowanceKey, jsonEncode(_state!.toJson()));
  }
}

File: emappapp/emappa-mobile/src/services/allowance.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { AllowanceState } from '../types';

const STORAGE_KEY = 'allowance_state_v1';

class AllowanceService {
  private _state: AllowanceState | null = null;

  get state(): AllowanceState | null {
    return this._state;
  }

  async load(): Promise<void> {
    try {
      const raw = await AsyncStorage.getItem(STORAGE_KEY);
      if (raw) {
        this._state = JSON.parse(raw) as AllowanceState;
      } else {
        // Sensible defaults for a new user
        const now = new Date();
        const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
        this._state = {
          monthlyAllowanceKwh: 100,
          remainingKwh: 100,
          rollover: true,
          lastReset: startOfMonth.toISOString(),
        };
        await this.save();
      }
    } catch (e) {
      console.error('Failed to load allowance:', e);
    }
  }

  private async save(): Promise<void> {
    if (!this._state) return;
    try {
      await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(this._state));
    } catch (e) {
      console.error('Failed to save allowance:', e);
    }
  }

  // If we've crossed into a new month, reset period
  async ensureCurrentPeriod(): Promise<void> {
    if (!this._state) return;

    const now = new Date();
    const currentPeriod = new Date(now.getFullYear(), now.getMonth(), 1);
    const lastReset = new Date(this._state.lastReset);
    const lastPeriod = new Date(lastReset.getFullYear(), lastReset.getMonth(), 1);

    if (currentPeriod > lastPeriod) {
      // New month
      let newRemaining = this._state.monthlyAllowanceKwh;
      if (this._state.rollover) {
        newRemaining += this._state.remainingKwh; // carryover unused
      }
      this._state = {
        ...this._state,
        remainingKwh: newRemaining,
        lastReset: currentPeriod.toISOString(),
      };
      await this.save();
    }
  }

  // Update plan settings
  async setPlan(
    monthlyAllowanceKwh: number,
    rollover: boolean,
    newPlan: boolean = false
  ): Promise<void> {
    if (!this._state) await this.load();
    if (!this._state) return;

    // If selecting a brand-new plan, reset remaining to the plan amount
    const newRemaining = newPlan
      ? monthlyAllowanceKwh
      : this._state.remainingKwh;

    this._state = {
      ...this._state,
      monthlyAllowanceKwh,
      rollover,
      remainingKwh: newRemaining,
    };
    await this.save();
  }

  // Deduct used allowance. Returns how much was actually deducted.
  async consume(kwh: number): Promise<number> {
    if (!this._state) await this.load();
    if (!this._state) return 0;

    const deduct = Math.min(Math.max(kwh, 0), this._state.remainingKwh);
    this._state = {
      ...this._state,
      remainingKwh: this._state.remainingKwh - deduct,
    };
    await this.save();
    return deduct;
  }
}

// Singleton instance
export const allowanceService = new AllowanceService();
